Passing a file to `FileArchive` causes engine panic

I did the following:

```
const lambda = new aws.lambda.Function("myfunction", {
    code: new pulumi.asset.FileArchive("./app.js"), // point to file, not directory
    role: role.arn,
    handler: "app.handler",
    runtime: aws.lambda.NodeJS8d10Runtime,
});
```

Output:

```
$ pulumi update
Previewing update of stack 'pulumi-serverless-dev'
Previewing changes:

     Type                              Name                                     Plan       Info
 +   pulumi:pulumi:Stack               pulumi-serverless-pulumi-serverless-dev  create
 +   ├─ aws-serverless:apigateway:API  myapi                                    create
 +   ├─ aws:iam:Role                   mylambda-role                            create
 +   ├─ aws:dynamodb:Table             counterTable                             create
 +   └─ aws:iam:RolePolicyAttachment   mylambda-access                          create
panic: runtime error: invalid memory address or nil pointer dereference
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x11e83a2]

goroutine 215 [running]:
github.com/pulumi/pulumi/pkg/util/contract.IgnoreClose(0x0, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/util/contract/ignore.go:33 +0x22
panic(0x194d360, 0x2027720)
        /Users/travis/.gimme/versions/go1.9.darwin.amd64/src/runtime/panic.go:491 +0x283
github.com/pulumi/pulumi/pkg/resource.addNextFileToTar(0x0, 0x0, 0xc42051d680, 0x0, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/asset.go:900 +0x52
github.com/pulumi/pulumi/pkg/resource.(*Archive).archiveTar(0xc4205379f0, 0x2588270, 0xc4200a7680, 0x0, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/asset.go:932 +0x11f
github.com/pulumi/pulumi/pkg/resource.(*Archive).Archive(0xc4205379f0, 0x1, 0x2588270, 0xc4200a7680, 0xc4200a7680, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/asset.go:886 +0x24c
github.com/pulumi/pulumi/pkg/resource.(*Archive).EnsureHash(0xc4205379f0, 0x0, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/asset.go:1019 +0x114
github.com/pulumi/pulumi/pkg/resource/plugin.UnmarshalPropertyValue(0xc420553d50, 0xc420448730, 0x49, 0x100000100, 0xc420083e98, 0xc4200bad40, 0x2)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/plugin/rpc.go:287 +0x5f1
github.com/pulumi/pulumi/pkg/resource/plugin.UnmarshalProperties(0xc4207401a0, 0xc420448730, 0x49, 0x100000100, 0x1, 0xc420553eb0, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/plugin/rpc.go:200 +0x27c
github.com/pulumi/pulumi/pkg/resource/deploy.(*resmon).RegisterResource(0xc4200baf80, 0x1fd7c80, 0xc4206a6c60, 0xc42027f0e0, 0xc4200baf80, 0x1, 0x1)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/deploy/source_eval.go:357 +0x3d1
github.com/pulumi/pulumi/sdk/proto/go._ResourceMonitor_RegisterResource_Handler.func1(0x1fd7c80, 0xc4206a6c60, 0x1a064e0, 0xc42027f0e0, 0x1fd7c80, 0xc4206a6c60, 0x1fdd2e0,0x205f8c0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/sdk/proto/go/resource.pb.go:368 +0x86
github.com/pulumi/pulumi/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x1fd7c80, 0xc4206a6c60, 0x1a064e0, 0xc42027f0e0, 0xc4205494c0, 0xc4205494e0, 0x0, 0x0, 0x1fc8ac0, 0xc420047050)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:61 +0x326
github.com/pulumi/pulumi/sdk/proto/go._ResourceMonitor_RegisterResource_Handler(0x19b35a0, 0xc4200baf80, 0x1fd7c80, 0xc4206a6bd0, 0xc42019cd20, 0xc420549140, 0x0, 0x0, 0x0, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/sdk/proto/go/resource.pb.go:370 +0x16d
github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4204629a0, 0x1fdb480, 0xc42050d680, 0xc420433680, 0xc42041e4b0, 0x202fc90, 0x0, 0x0, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:923 +0x92d
github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4204629a0, 0x1fdb480, 0xc42050d680, 0xc420433680, 0x0)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:1148 +0x1528
github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc420282270, 0xc4204629a0, 0x1fdb480, 0xc42050d680, 0xc420433680)
        /Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:637 +0x9f
created by github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
        /Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:635 +0xa1
```

--- a/pkg/resource/asset.go
+++ b/pkg/resource/asset.go
@@ -291,7 +291,7 @@ func (a *Asset) Read() (*Blob, error) {
 	} else if a.IsURI() {
 		return a.readURI()
 	}
-	return nil, nil
+	return nil, errors.New("unrecognized asset type")
 }
 
 func (a *Asset) readText() (*Blob, error) {

